### PR TITLE
feat: Add Schema Directives to SDL

### LINKED ISSUE #947: Support for operation directives
The discussion on issue #53 seems to be about schema directives, so I'm opening a new issue regarding query directives which is also not supported, 

It seems quite simple to add this feature by passing an extra field to the `GraphQLSchema` constructor here:
https://github.com/graphql-nexus/nexus/blob/8e0ce1b4a156cd50b061b47c8a1241729b4dd2a3/src/builder.ts#L1700-L1709

Since that is just one function down from `makeSchema` and uses the same type, all it requires is adding a new field on the `SchemaConfig` [here](https://github.com/graphql-nexus/nexus/blob/8e0ce1b4a156cd50b061b47c8a1241729b4dd2a3/src/builder.ts#L259-L277) and passing it along. I tested it locally, and it works perfectly. The schema gets generated correctly with the directive specification at the top, and the information is passed to the resolvers at `info.fieldNodes.directives`  as [graphql-js defines](https://github.com/graphql/graphql-js/blob/6a5f51f3bd07f3914c7f1ec01908e9e325b77d58/src/language/ast.ts#L304).

Here is how it ended up being with this change:
```typescript
/* file: directives.ts */
import {GraphQLDirective, GraphQLInt} from 'graphql';

export const cacheDirective = new GraphQLDirective({
    name: 'cached',
    args: {
        lifetime: {
            type: GraphQLInt,
        },
    },
    locations: ['FIELD'],
});

/* file: schema.ts */
import {makeSchema} from 'nexus';
import {cacheDirective} from './directives';

export const NexusGraphQLSchema = makeSchema({
    directives: [cacheDirective],
    ...
});

/* file: testResolver.ts */
import {queryField} from 'nexus';
import {cacheDirective} from '../directives';
import {getArgumentValues} from 'graphql/execution/values';

export const testField = queryField(def => def.string('testField', {
     resolve: (_, __, ___, info) => {
        const cacheConfiguration = getDirectiveValues(cacheDirective, info.fieldNodes[0], info.variableValues);
        return cacheConfiguration?.lifetime?.toFixed(0) ?? 'no-cache';
    },
});
```

It seems deceptively simple for no one to have done it by now, is there a problem with this that happens not to affect my application, or should I open a PR?

### PR DESCRIPTION
**Pull request by tgriesser**
Closes: #947, #53

Helps with: #148

WIP: Still needs docs / additional tests / cleanup of a bit of the builder layer / fixing types

Adds a `directives` property to all definition props which takes an array of directives. Adds them to an `astNode` which are then printed via a custom schema printer.

Definition:

```ts
const someDirective = directive({
  name: 'SomeDirective',
  locations: ['OBJECT'],
  args: {
    bool: nonNull(booleanArg())
  }
})
```

Use:

```ts
objectType({
  name: 'SomeType',
  directives: [someDirective({bool: true})]
})
```

or 

```ts
import { addDirective } from 'nexus'

objectType({
  name: 'SomeType',
  directives: [addDirective('SomeDirective', {bool: true})]
})
```

Prints:

```graphql
type SomeType @SomeDirective(bool: true) {
  # ... fields
}
```

See changes in the [`kitchenSink/__app.ts`](https://github.com/graphql-nexus/nexus/compare/tgriesser/feat/sdl-directives?expand=1#diff-4459f76e829acc4ad1fa32dc0267afba074f726199c749ef0be08ce3d2fa4fe7) for more example uses.

### CODE REVIEW

**Comment by Sytten** [comment]
Dont close #148, still needs a plugin for it to make is nice for people to use (should be trivial once this is merged)

File: src/printSchemaWithDirectives.ts
Comment (Sytten) [inline src/printSchemaWithDirectives.ts:36]: I don't think we need a custom printer, all graphql versions that nexus support have a schema printer that does print the directives:
https://github.com/graphql/graphql-js/blob/v14.5.8/src/utilities/schemaPrinter.js
https://github.com/graphql/graphql-js/blob/v15.5.1/src/utilities/printSchema.js
And they should be exported by `graphql/utilities`.

File: src/makeSchema.ts
Comment (Sytten) [inline src/makeSchema.ts:81]: Do we want to export that anymore?

File: src/makeSchema.ts
Comment (Sytten) [inline src/makeSchema.ts:119]: I have a TS error in editor for that line `Property 'operationTypes' is missing in type 'ScalarTypeDefinitionNode' but required in type 'SchemaDefinitionNode'`

Comment (Sytten) [reply src/makeSchema.ts:119]: I feel this is misleading for future changes, this returns an AST node. If some would need to override another property in the `astNode` they might override the `astNode` set by `schemaDirectives`.

File: src/builder.ts
Comment (Sytten) [inline src/builder.ts:934]: I said it in the `makeSchema`, but I want to drive the point, since this adds a whole `astNode` I feel that someone that might want to modify the AST node for another purpose will get blocked by this code or might override it entirely. I think we should at the very least wrap that in an array and a merge function or even better if only the `directives` field of the node was returned by `maybeAddDirectiveUses` (free and member functions) and the `astNode` was generated by another method.

**Review by Sytten — CHANGES_REQUESTED** [review]

File: src/definitions/args.ts
Comment (tgriesser) [reply src/definitions/args.ts:57]: Yeah I had initially tried this api as:
```ts
directives(t) {
  t.MyDirective(args)
  t.OtherDirective(args)
}
```

Where the members were dynamically based on directive declarations & where they can be positioned. It generally worked, but the types were pretty gnarly & seemed to be a lot of complexity for something that could just be a simple runtime error.

File: src/definitions/directive.ts
Comment (tgriesser) [reply src/definitions/directive.ts:193]: Hmm... Need to check on that

File: src/definitions/directive.ts
Comment (tgriesser) [reply src/definitions/directive.ts:71]: No, good catch - need to do that

File: src/printSchemaWithDirectives.ts
Comment (tgriesser) [reply src/printSchemaWithDirectives.ts:36]: I believe it only prints the directives declarations, not the directive uses on the SDL types/fields

File: src/makeSchema.ts
Comment (tgriesser) [reply src/makeSchema.ts:81]: Yeah I think I'm using it in the tests, I don't believe it's publicly exported on the module

File: src/printSchemaWithDirectives.ts
Comment (Sytten) [reply src/printSchemaWithDirectives.ts:36]: Good point, bit annoying that we have to copy paste/maintain all that code. We should reference the commit it was taken from and maybe the diff vs the original to reduce upgrade burden.

Comment (tgriesser) [reply src/printSchemaWithDirectives.ts:36]: Yeah, I'm not too worried about it in the sense that since it's just printing code the paths are pretty simple and easy to validate correctness of (not a lot of business logic that you'd otherwise be worried about with copy/paste).

File: src/makeSchema.ts
Comment (tgriesser) [reply src/makeSchema.ts:119]: Yeah agreed, meant to come back to this / out this as a TODO

File: src/builder.ts
Comment (tgriesser) [reply src/builder.ts:934]: > I feel that someone that might want to modify the AST node for another purpose will get blocked by this code or might override it entirely

Yeah, fair points - I had considered that and was thinking of just crossing that bridge if/when it came to it, since `astNode` isn't part of the public api for any of the nexus builder fn's, I wasn't sure if that would actually happen in practice.

Comment (Sytten) [reply src/builder.ts:934]: In theory no, we should never touch the AST node since it is supposed to the result of the parsing of the schema.
I still think it should be explicit in the code that we are converting the directives to an AST node. Something like:
```typescript
this.createAstNode([this.maybeAddDirectiveUses])
```
I suspect that a future release of graphql will add backend directives in the spec, so it will just be easier to move to that then.

File: src/definitions/directive.ts
Comment (tgriesser) [reply src/definitions/directive.ts:96]: Actually recently found that we don't do this for field names, but we should 😬

**Comment by tgriesser** [comment]
> Is there an estimated time for this new feature? 👀

Going to probably work on it a bit today/tomorrow.

**Comment by davidsoderberg** [comment]
Any progress on this one?

**Comment by yasaichi** [comment]
What's a blocker of merging this PR?

**Comment by victoriris** [comment]
@davidsoderberg & @yasaichi just to let you know... I needed this for Apollo Federation support and ended up switching to type-graphql a month ago

**Comment by davidsoderberg** [comment]
I also ended up switching to type-graphql... :(

File: src/builder.ts
Comment (redbar0n) [reply src/builder.ts:965]: @tgriesser what would be the shortest path to unblock this pull request so it can go through?

**Comment by zingerj** [comment]
Hey all, what's the latest on getting this merged? Looks like there's one failing test; is that it to get this across the line? Thanks for your work here!

File: src/builder.ts
Comment (redbar0n) [reply src/builder.ts:965]: @zingerj this is the blocker.

**Comment by YElyousfi** [comment]
When can we expect this to get merged? Federation relies on directives.

**Comment by meabed** [comment]
This is great feature - thank you!! 
Would it be ready to merge soon?

**Comment by yasaichi** [comment]
To those who can't wait this PR merged
You have the following alternative libraries supporting directives:

- [TypeGraphQL](https://typegraphql.com/docs/directives.html) (mentioned above)
- [Pothos](https://pothos-graphql.dev/docs/plugins/directives) (formerly GiraphQL)

**Comment by BahaaZidan** [comment]
Please merge this!

File: src/builder.ts
Comment (tgriesser) [reply src/builder.ts:1054]: Going to punt on this for now - going to assume that if you need to modify the ASTNode you can use something else more specialized like graphql-tools

**Comment by tgriesser** [comment]
This is merged, and is now released under `nexus@1.4.0-next.7`, no docs yet other than in the [PR description](https://github.com/graphql-nexus/nexus/pull/952#issue-949724569). 

No official plugin for federation, but the primitives are there - feel free to open a PR if you make progress on something generic and would like to add a plugin to the core.

Additionally, please give any feedback or let me know if you run into issues with it, will get this into an official `1.4.0` soon

**Comment by meabed** [comment]
@tgriesser - Thank you! any date we expecting this to be released?

**Comment by cqhung1412** [comment]
@tgriesser - Please release this soon with the docs.

**Comment by unrevised6419** [comment]
Hey folks! Thanks for working on this feature for Nexus!

Should there be any type of completion for the directive args when using it? 🤔 

image

image

I've augmented it using this

```ts
declare module 'nexus/dist/core' {
  export interface NexusDirectiveDef {
    (arg: CacheHint): NexusDirectiveUse<"cacheControl">;
  }
}
```

PS: Does anyone know when a new minor release will be done with this feature?

Not related: I'm using this feature to implement cache control in Strapi. I started with using resolvers, but I think the directive approach is much better.

https://github.com/strapi/strapi/pull/20254

**Pull request merged**
